### PR TITLE
Patch missing footpath in scenarios

### LINKED ISSUE #22972: Missing brown dirt path in Six Flags Magic Mountain scenarios
### Operating System

Windows 10, 64-bit

### OpenRCT2 build

OpenRCT2, v0.4.15 (af7a2de72c on develop) provided by GitHub

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

The bug also exists in the original game (RCT2/RCTC)

### Describe the issue

There is a brown dirt path missing in both the "Six Flags Magic Mountain" and "Build your own Six Flags Magic Mountain" scenario. The missing path causes guests to get lost. This issue is present in vanilla RCT2 as well.

### Steps to reproduce

None needed, you can always view this issue by loading the aforementioned scenarios and looking in the middle of the map near the ride: "The Escape".

### Attachments

OpenRCT2:

Image

Vanilla RCT2:

Image

### PR DESCRIPTION
**Pull request by staticinvocation**
This adds path patching to scenario patching to fix #22972. The missing path was fixed in the Six Flags Magic Mountain scenarios.

### CODE REVIEW

**Comment by tupaschoal** [comment]
> Currently this only adds the ability to add flat footpaths and they will connect to each nearby footpath, correct?
> 
> For other use-cases it would be nice to see the other options, that are supported in the tile inspector, being added.

I haven't looked at the code yet, but let's keep in mind that the scenario patcher is being built on a per need basis, we don't want to support options without evidence that they will help fixing scenario bugs.

**Comment by ZeeMaji** [comment]
Can this place sloped footpath pieces? That will be needed for fixing some errors in other scenarios (eg: Alton Towers)

**Comment by staticinvocation** [comment]
> Can this place sloped footpath pieces? That will be needed for fixing some errors in other scenarios (eg: Alton Towers)

Okay I see #20070 and no, this cannot currently place sloped pieces. However that should be an easy add and I can fix all of these.

**Comment by HtotheTML** [comment]
> Okay I see #20070 and no, this cannot currently place sloped pieces. However that should be an easy add and I can fix all of these.

Does this work on underground path errors like those in Mines of Africa and Rollercoaster Heaven?

**Comment by staticinvocation** [comment]
> Does this work on underground path errors like those in Mines of Africa and Rollercoaster Heaven?

Yes, I just committed a patch for the latter: 
image

**Comment by Gymnasiast** [comment]
The code looks reasonable to me.

What I would like to see here:
- Confirmation from @ZeeMaji and/or @HtotheTML that it addresses their points.
- A changelog entry. One entry is sufficient. Include the issue number for the path fixes that have their own issue. If one of them does not have its own issue, also add the number of this PR.
- An approval from @tupaschoal

**Comment by tupaschoal** [comment]
I've been a bit busy, but I'll try to get it done today

**Comment by ZeeMaji** [comment]
> * Confirmation from @ZeeMaji and/or HtotheTML that it addresses their points.

The missing sloped path in alton towers is still missing, the missing piece inside the fortress has been fixed though.
Screenshot at 2024-11-07 18-09-54

**Comment by tupaschoal** [comment]
Here's a summary of what I've done:
- Added `.parkpatch` fixes to the corresponding `.sea` scenarios touched
- Reorganized the commit history (including commit titles) to be:
  1. Fix for flat footpathes
  2. Fix for sloped footpaths
  3. Changelog and network bump (I decided to bump the network version since I think the new footpaths will cause desyncs, as they are amending broken footpaths)
- Fixed the `getDirection` issue I raised.
- Made coordinate formatting consistent with other `.parkpatch` files
- Fixed the missing footpath on "Alton Towers" pointed out by @ZeeMaji 
- Tested all current modifications are sane and needed

This would be ok to be merged as is. Note that it fully fixes #22972, but not #22070. For the latter to be fixed, we'd still need to:

- [ ] Fix Iceberg Islands queue: This one is weird, there is no bug on the `.sc4`, where all queues have the blue footpath, but the `.sea` one has some fully converted to yellow and others (like the bobsled) partly yellow, partly blue. I wonder if this can be a bug in importing. Thoughts @Gymnasiast ?
- [ ] Blackpool pleasure beach mismatched railing near Drinks Stall 2: this one is weird, it seems like we dropped the railings at some point. Would also like to hear @Gymnasiast opinion
- [ ] Mines of Africa incomplete queue for Diamond Ride (both `.sea` and `.sc6`)
- [ ] Mirage Madness missing footpath leading to monorail 1 (only `.sc6`)
- [ ] Rollercoaster Heaven missing path to drinks stall 2 (both `.sea` and `.sc6`)

**Comment by ZeeMaji** [comment]
The BPB railings near drink stall 2 were actually an inaccuracy, for some reason the paths in that area are actually road instead of gray tarmac, which at some point in the past was imported as the RCT2 road surface which rendered railings on slopes unlike other RCT1 path surfaces. Semi-recently this was improved by implementing a proper RCT1 road surface which does not render railings on slopes like other RCT1 path surfaces. So that area presently looks correct and does not need any alterations.

**Comment by ZeeMaji** [comment]
Iceberg islands is broken in SC4 too, but it's more subtle. It uses blue queue throughout but the railing styles are mismatched, it's supposed to use the concrete railings throughout the whole queue but the tiles near the entrance use wooden railings. See: https://user-images.githubusercontent.com/108596959/235015023-c5d8aee3-1583-4d24-bc87-83dbbf1984e3.png

**Comment by staticinvocation** [comment]
Thank you @tupaschoal 😄

**Comment by tupaschoal** [comment]
I have it all but the Iceberg Islands fixed pushed. Iceberg Islands `.sc4` is done, but since the `.sea` requires rebuilding the whole queues to be blue, I still need a bit more time.

**Comment by ZeeMaji** [comment]
For iceberg islands .sea, i would suggest just replacing the couple of blue tiles with yellow tiles (along with concrete railings on them) as i feel like that's how it would have been imported into RCTC if the original error in the SC4 didn't exist due to the more limited path system in vanilla/RCTC. Vanilla didn't let you mix surfaces/railings, the way the error is set up suggests that queues were imported by railing type as the parts with concrete railings in RCT1 got converted to the yellow queue with concrete railings, while the parts with wooden railings got converted to blue queue with wooden railings.
My proposed fix would be this essentially.
Screenshot at 2024-11-14 11-34-12

**Comment by tupaschoal** [comment]
I thought about this too, will let others voice their opinion.

In any case, your previous comment made me notice that the railings for the mini golf queue were also wrong

**Comment by ZeeMaji** [comment]
I'm a bit hesitant to say if the mini golf queue is wrong, it's fully wood throughout so it feels a lot more deliberate unlike the bobsled queue where it definitely seems more like an error due to the mixed railings.

A similar thing exists in a few other LL scenarios i believe, Nevermore park comes to mind as three queues there use concrete railings while one uses wooden railings.

**Comment by tupaschoal** [comment]
> I'm a bit hesitant to say if the mini golf queue is wrong, it's fully wood throughout so it feels a lot more deliberate unlike the bobsled queue where it definitely seems more like an error due to the mixed railings.

I get your point, it's definitely less clear, but at the same time all other 3 queues have concrete railings, but the mini-golf, so it seems weird.

**Comment by tupaschoal** [comment]
I'm done with all changes in this review. It fixes everything stated in the two issues. Note that for the Iceberg Islands I went with all yellow queues for `.sea` and all blue for `sc4`. I did remake the mini-golf course for both to use the concrete railings.

**Comment by Gymnasiast** [comment]
I’ll take a look at the Iceberg Islands mini golf queue. I hope to get reviews in this weekend.

**Comment by tupaschoal** [comment]
I've rebased to fix changelog and network conflicts.

**Comment by staticinvocation** [comment]
@Gymnasiast if you have some time could you please take a look at this? I get extra credit on a school assignment if this gets merged in a few days 😃

**Comment by janisozaur** [comment]
We would like to merge it this week, for upcoming 0.4.17 release, we just need to review it

**Comment by janisozaur** [comment]
@staticinvocation what's your deadline?

**Comment by staticinvocation** [comment]
> @staticinvocation what's your deadline?

Monday the 9th. Thanks

File: src/openrct2/rct12/ScenarioPatcher.cpp [hunk 1]
```diff
@@ -93,7 +104,33 @@ static u8string ToOwnershipJsonKey(int ownershipType)
...
+
+    out.emplace_back(
+        OpenRCT2::Json::GetNumber<int32_t>(coordinatesArray[0]), OpenRCT2::Json::GetNumber<int32_t>(coordinatesArray[1]));
+}
+
+static void readCoordinate(std::vector<TileCoordsXYZ>& out, const json_t& coordinatesArray)
+{
+    if (coordinatesArray.size() != 3)
+    {
+        OpenRCT2::Guard::Assert(0, "Fix coordinates sub array should have 3 elements");
```
Comment (janisozaur) [inline src/openrct2/rct12/ScenarioPatcher.cpp:123]: It's a shame we use asserts for such conditions. There should be an error displayed with link to relevant source file and line - not part of this PR.

File: data/scenario_patches/2980c28.parkpatch [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (janisozaur) [inline data/scenario_patches/2980c28.parkpatch:1]: LGTM

File: data/scenario_patches/0b8cc95.parkpatch [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (janisozaur) [inline data/scenario_patches/0b8cc95.parkpatch:1]: LGTM

File: data/scenario_patches/2696a05.parkpatch [hunk 4]
```diff
@@ -18,5 +18,20 @@
                 [  140, 74 ], [  141, 74 ], [  142, 74 ], [ 143, 74 ], [ 144, 74 ], [ 145, 74 ], [ 146, 74 ], [ 147, 74 ]
             ]
         }
-    }
+    },
```
Comment (janisozaur) [inline data/scenario_patches/2696a05.parkpatch:21]: LGTM

File: data/scenario_patches/7ffdb44.parkpatch [hunk 5]
```diff
@@ -4,11 +4,11 @@
     "land_ownership": {
```
Comment (janisozaur) [inline data/scenario_patches/7ffdb44.parkpatch:4]: LGTM

File: data/scenario_patches/4a762ae.parkpatch [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (janisozaur) [inline data/scenario_patches/4a762ae.parkpatch:1]: LGTM

File: data/scenario_patches/5c95a4e.parkpatch [hunk 7]
```diff
@@ -7,5 +7,18 @@
                 [ 11, 31 ], [ 68, 112 ], [ 72, 118 ]
             ]
         }
-    }
+    },
```
Comment (janisozaur) [inline data/scenario_patches/5c95a4e.parkpatch:10]: LGTM

File: data/scenario_patches/c82272a.parkpatch [hunk 8]
```diff
@@ -21,5 +21,12 @@
                 [ 103, 76 ], [ 104, 76 ]
             ]
         }
+    ],
```
Comment (janisozaur) [inline data/scenario_patches/c82272a.parkpatch:24]: LGTM

**Review by janisozaur — APPROVED** [review]
I have tested all the files and they do indeed fix path/rails issues in all the places. Thanks a lot for the contribution @staticinvocation

**Comment by janisozaur** [comment]
@staticinvocation can you name what school / course it is? We are always happy to have students contribute code, you can recommend our project to others.

**Comment by staticinvocation** [comment]
@janisozaur [Sure](https://web.eecs.umich.edu/~weimerw/481/hw6.html)! Here are the assignment details on the course website. This project has actually been a favorite for the course going back several years and I’m aware of 2 other people in my class who also picked this project.

**Pull request merged**